x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/repositories/PutCcrRestoreSessionAction.java

```diff
@@ -0,0 +1,155 @@
+        return PutCcrRestoreSessionAction.PutCcrRestoreSessionResponse::new;
+    }
+
+    public static class TransportPutCcrRestoreSessionAction
+        extends TransportSingleShardAction<PutCcrRestoreSessionRequest, PutCcrRestoreSessionResponse> {
+
+        private final IndicesService indicesService;
+        private final CcrRestoreSourceService ccrRestoreService;
+
+        @Inject
+        public TransportPutCcrRestoreSessionAction(ThreadPool threadPool, ClusterService clusterService, ActionFilters actionFilters,
+                                                   IndexNameExpressionResolver resolver, TransportService transportService,
+                                                   IndicesService indicesService, CcrRestoreSourceService ccrRestoreService) {
+            super(NAME, threadPool, clusterService, transportService, actionFilters, resolver, PutCcrRestoreSessionRequest::new,
+                ThreadPool.Names.GENERIC);
+            this.indicesService = indicesService;
+            this.ccrRestoreService = ccrRestoreService;
+        }
+
+        @Override
+        protected PutCcrRestoreSessionResponse shardOperation(PutCcrRestoreSessionRequest request, ShardId shardId) throws IOException {
+            IndexShard indexShard = indicesService.getShardOrNull(shardId);
+            if (indexShard == null) {
+                throw new ShardNotFoundException(shardId);
+            }
+            Store.MetadataSnapshot sourceMetaData = ccrRestoreService.openSession(request.getSessionUUID(), indexShard);
+            Store.RecoveryDiff recoveryDiff = sourceMetaData.recoveryDiff(request.getMetaData());
+
+            ArrayList<StoreFileMetaData> filesToRecover = new ArrayList<>(recoveryDiff.different);
+            filesToRecover.addAll(recoveryDiff.missing);
+            return new PutCcrRestoreSessionResponse(indexShard.routingEntry().currentNodeId(), recoveryDiff.identical, filesToRecover);
+        }
+
+        @Override
+        protected PutCcrRestoreSessionResponse newResponse() {
+            return new PutCcrRestoreSessionResponse();
+        }
+
+        @Override
+        protected boolean resolveIndex(PutCcrRestoreSessionRequest request) {
+            return false;
+        }
+
+        @Override
+        protected ShardsIterator shards(ClusterState state, InternalRequest request) {
+            final ShardId shardId = request.request().getShardId();
+            // The index uuid is not correct if we restore with a rename
+            IndexShardRoutingTable shardRoutingTable = state.routingTable().shardRoutingTable(shardId.getIndexName(), shardId.id());
+            return shardRoutingTable.primaryShardIt();
+        }
+    }
+
+
+    public static class PutCcrRestoreSessionResponse extends ActionResponse {
+
+        private String nodeId;
+        private List<StoreFileMetaData> identicalFiles;
+        private List<StoreFileMetaData> filesToRecover;
```

Comment:
perhaps leave these classes empty for now as the fields are not being used / tested yet?

Reply:
Removed unused fields